### DISCUSSION TITLE
Clean example of a partitioned asset with Snowflake?

### DISCUSSION THREAD

**Discussion by clayheaton** [dagster-io/dagster#17507] [Q&A] [answered]
I've read the docs and poked around in the example projects and have yet to find what I would consider to be a simple and clean example of a simple Daily-partitioned SDA for working with Snowflake. Can somebody help me fill this out?

Main questions I have:

- How am I supposed to feed the partition date into the query?
- Do I need to do anything to support backfills/overwriting of partitions by Dagster?
- Do I need `DAY_KEY` as part of the `SELECT` statement or does Dagster create that column in the partitioned output table? If no, then do I need to add something to the DataFrame for Dagster to output the date column properly?

```python
from dagster import DailyPartitionsDefinition, asset, OpExecutionContext
from contextlib import closing
import pandas as pd

@asset(
    required_resource_keys={"MySnowflakeResource"},
    partitions_def=DailyPartitionsDefinition(start_date="2023-09-01")
)
def TEST_PARTITIONS(context: OpExecutionContext) -> pd.DataFrame:
    """
    Testing Dagster partitioning. I want to query a Snowflake table for a daily count
    and return the count to a partitioned Snowflake table, TEST_PARTITIONS, that is managed by Dagster.
    """

    # DAY_KEY is a date field in the format YYYY-MM-DD
    query = """
        SELECT DAY_KEY, COUNT(DISTINCT TRANSACTION_ID) AS TRANSACTION_COUNT
        FROM MY_DATABASE.MY_SCHEMA.MY_TABLE
        WHERE  DAY_KEY = { ???? }
        GROUP BY 1
        ORDER BY 1
    """

    resource = getattr(context.resources, "MySnowflakeResource")
    with resource.get_connection() as conn:
        with closing(conn.cursor()) as cursor:
            cursor.execute(query)
            df = cursor.fetch_pandas_all()
            return df
```

**Comment by marcilj** [comment, marked as answer]
> How am I supposed to feed the partition date into the query?

The partition is provided through the `context`.
There's multiple function that allow you to access the provided partition key.
![image](https://github.com/dagster-io/dagster/assets/89793010/90cfcaf4-58b4-4055-a3f8-f04de1ec4fa9)

The input partition represent the partition of the parent asset, and the output partition of the asset you want to create.

In your case you could use something like `date = context.asset_partition_key_for_output()`

> Do I need to do anything to support backfills/overwriting of partitions by Dagster?

Your `IO_MANAGER` should handle that for you.
When you `return` a dataframe, the dataframe is sent to the configured `IO_MANAGER` of the asset that handle the storage and the retrieval of the data.

In your case if you want to store data in a table in Snowflake, you can use the provided io_manager created by dagster or create your own.

If you choose to use the provided one, you can configure it using the following code.

```python
from dagster_snowflake import build_snowflake_io_manager, snowflake_resource

snowflake_io_manager = build_snowflake_io_manager(
    [SnowflakePandasTypeHandler()]
)

SHARED_SNOWFLAKE_CONF = {
    "account": os.getenv("SNOWFLAKE_ACCOUNT", ""),
    "user": os.getenv("SNOWFLAKE_USER", ""),
    "password": os.getenv("SNOWFLAKE_PASSWORD", ""),
    "warehouse": os.getenv("SNOWFLAKE_WAREHOUSE", ""),
    "database": os.getenv("SNOWFLAKE_DATABASE", ""),
    "role": os.getenv("SNOWFLAKE_ROLE", ""),
}

RESOURCES_SANDBOX = {
    "warehouse_io_manager": snowflake_io_manager.configured(
        dict(**SHARED_SNOWFLAKE_CONF)
    )
}
```

If you use the default Snowflake io manager, it will by default `delete` the data from the selected partition, then `insert` the content of your dataframe. That way you can be sure that you only have fresh data in your table.

If you want to create yours, then you can do that by creating a class that Inherit from `ConfigurableIOManager` and configured the `def handle_output()` to store data, and `def load_input()` to retrieve data. All the backfill logic should be included in this IO_Manager if you decide to create it yourself. https://docs.dagster.io/concepts/io-management/io-managers#defining-an-io-manager

To use an IO Manager with your asset you need to configure it in your asset `decorator` like this.
```python
@asset(
    required_resource_keys={"MySnowflakeResource"},
    partitions_def=DailyPartitionsDefinition(start_date="2023-09-01"),
    io_manager_key="warehouse_io_manager",
)
def TEST_PARTITIONS(context: OpExecutionContext) -> pd.DataFrame:
```

The name of the ressource is the name of the key in `RESOURCES_SANDBOX` in this example.


> Do I need DAY_KEY as part of the SELECT statement or does Dagster create that column in the partitioned output table? If no, then do I need to add something to the DataFrame for Dagster to output the date column properly?

Dagster doesn't add any columns that is not part of the dataframe, it only store the dataframe itself as a table, or insert new row in the created table.

In your example, the final table will have 2 columns, 1 called `DAY_KEY` and 1 called `TRANSACTION_COUNT`. Since your `DAY_KEY` column represent your `Partition` you don't have to add any new one and it should work as expected.

If you have a configured `key_prefix` in your `asset` decorator, Snowflake IO Manager will take that as your destination schema name. This is very usefull if you want to use the same ressource for multiple assets stored in different schema.

That would look like this
```python
@asset(
    required_resource_keys={"MySnowflakeResource"},
    partitions_def=DailyPartitionsDefinition(start_date="2023-09-01"),
    io_manager_key="warehouse_io_manager",
    key_prefix="my_destination_schema",
)
def TEST_PARTITIONS(context: OpExecutionContext) -> pd.DataFrame:
```

Then, you need to tell snowflake which column is the one where you store your `partition key`. If not it couldn't possibly know which set of rows you want to update if you do a backfill.

To do that you need to configure a metadata key called `partition_expr` with the value as the name of your column.

```python
@asset(
    required_resource_keys={"MySnowflakeResource"},
    partitions_def=DailyPartitionsDefinition(start_date="2023-09-01"),
    io_manager_key="warehouse_io_manager",
    key_prefix="my_destination_schema",
    metadata={
        "partition_expr": "DAY_KEY",
    },
)
def TEST_PARTITIONS(context: OpExecutionContext) -> pd.DataFrame:
```

If we put everything together that would be your final code

```python
from dagster import DailyPartitionsDefinition, asset, OpExecutionContext
from contextlib import closing
import pandas as pd

@asset(
    required_resource_keys={"MySnowflakeResource"},
    partitions_def=DailyPartitionsDefinition(start_date="2023-09-01"),
    io_manager_key="warehouse_io_manager",
    key_prefix="my_destination_schema",
    metadata={
        "partition_expr": "DAY_KEY",
    },
)
def TEST_PARTITIONS(context: OpExecutionContext) -> pd.DataFrame:
    """
    Testing Dagster partitioning. I want to query a Snowflake table for a daily count
    and return the count to a partitioned Snowflake table, TEST_PARTITIONS, that is managed by Dagster.
    """

    date = context.asset_partition_key_for_output()

    # DAY_KEY is a date field in the format YYYY-MM-DD
    query = f"""
        SELECT DAY_KEY, COUNT(DISTINCT TRANSACTION_ID) AS TRANSACTION_COUNT
        FROM MY_DATABASE.MY_SCHEMA.MY_TABLE
        WHERE  DAY_KEY = '{date}'
        GROUP BY 1
        ORDER BY 1
    """

    resource = getattr(context.resources, "MySnowflakeResource")
    with resource.get_connection() as conn:
        with closing(conn.cursor()) as cursor:
            cursor.execute(query)
            df = cursor.fetch_pandas_all()
            return df

```

With all of those thing you should be good to store partitioned assets in Snowflake.


In terms of structure, you should look at the `Fully Feature Project` to help you.
https://github.com/dagster-io/dagster/blob/master/examples/project_fully_featured/README.md

This will show you how you can configure different ressource per environment. https://github.com/dagster-io/dagster/blob/master/examples/project_fully_featured/project_fully_featured/resources/__init__.py

And how to define your current environment. https://github.com/dagster-io/dagster/blob/master/examples/project_fully_featured/project_fully_featured/__init__.py

I hope that helps a little.

**Reply by clayheaton** [reply]
Thank you very much. Your answer is very helpful.

When I started using Dagster, I ran into a problem with the default Snowflake IO manager because the way it connected to the database/schema conflicted with the RBAC in use at our company. Essentially, I need to run `USE SECONDARY ROLES ALL` before I run any other query if I want to have read access to the source tables/views. In other apps/libraries (DBeaver is the first that comes to mind), I have this set as a Bootstrap query that runs every time a database connection is made. Likewise, we have in place policy that requires us to use Snowflake's query tags. Hence, I was not connecting with the "automagic" Snowflake IO manager. 

I think my best move is to write my own Snowflake IO manager so that I can have it handle the bootstrap queries for setting roles and query tags while also being able to handle deletes from the partitioned table. Let me know if that seems wrong to you. Otherwise, thanks again for the detailed answer!
